In [1]:
# Import Pandas
# To work with DataFrames we use Pandas
# We use pandas for data manipulation, data cleaning and data analysis
# Pandas is utilised to format the data structure. It means it can convert any data structure into a Pandas DataFrame
import pandas as pd

# Import Numpy
# To find Statistical Summaries we use Numpy
import numpy as np

# Import subpackage of matplotlib 
import matplotlib.pyplot as plt

# Import Seaborn
import seaborn as sns


C:\Users\Sap\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


### Load the dataset

In [2]:
# Load the data and name it as df_house_price
df_house_price = pd.read_csv('House_Price1.csv')

#display the first five observations using head()
df_house_price.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200      None        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

## Data Overview 

In [3]:
# Name the Coloumns
df_house_price.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

- `price` : Price of the House. 
- `resid_area` : Resedential Area where my house is located
- `air_qual` : Air Quality near the house
- ``
- ``
  

In [4]:
# Basic Info of the Dataset
df_house_price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    506 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**
 - We have 13 Numerical columns & 3 Categorical Columns
 - We have null values in the data
 - We have total 16 columns in the data
  

In [5]:
# Shape of the data
df_house_price.shape

(506, 16)

**Interpretations** 
- We have total 16 columns 
- We have 506 Observations

**Interpretation**
 - We have outliers in the data , most of the outliers are in age coloumn

## Data Preprocessing

### A . Data Cleaning

In [6]:
# Show the First 5 Observations
df_house_price.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200      None        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [7]:
# Rename the Columns
# Step 1 : Show the Column Names
df_house_price.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [8]:
# Step 2 : Allocate ew column names to Old Column Names
# Make the Dictionary {'Old Column Name' : 'New Column Name'}
d1 = {'price' : 'Price',
      'resid_area' : 'Residential Area',
      'air_qual' : 'Air Quality',
      'room_num' : 'Room Numbers',
      'age' : 'Age',
      'distance' : 'Distance',
      'teachers' : 'Teachers',
      'poor_prop' : 'Proportion of Poor Population',
      'airport' : 'Airport',
      'n_hos_beds' : 'Number of Hospital Beds',
      'n_hot_rooms' : 'Number of Hotel Rooms',
      'waterbody' : 'Waterbody',
      'rainfall' : 'Rainfall',
      'bus_ter' : 'Bus Terminal',
      'parks' : 'Parks',
      'Sold' : 'Sold'}


In [9]:
df_house_price = df_house_price.rename(columns = d1)

In [10]:
# Authenticate the Results
df_house_price.columns

Index(['Price', 'Residential Area', 'Air Quality', 'Room Numbers', 'Age',
       'Distance', 'Teachers', 'Proportion of Poor Population', 'Airport',
       'Number of Hospital Beds', 'Number of Hotel Rooms', 'Waterbody',
       'Rainfall', 'Bus Terminal', 'Parks', 'Sold'],
      dtype='object')

### B. Null Value Handling

```
Laws that we need to follow while processing the null values (Specially for Manufacturing)
 - When we have less than 15% data missing -> Remove the observation
 - Between 20% to 70% data missing -> Impute the null values
 - Greater than 75% Values missing -> Remove the Column
```

In [11]:
# Find the Valid Null Values
df_house_price.isnull().sum()

Price                            0
Residential Area                 0
Air Quality                      0
Room Numbers                     0
Age                              0
Distance                         0
Teachers                         0
Proportion of Poor Population    0
Airport                          0
Number of Hospital Beds          8
Number of Hotel Rooms            0
Waterbody                        0
Rainfall                         0
Bus Terminal                     0
Parks                            0
Sold                             0
dtype: int64

In [12]:
# Percentage of Missing Values
round(df_house_price.isnull().sum() / df_house_price.shape[0] * 100),3

(Price                            0.0
 Residential Area                 0.0
 Air Quality                      0.0
 Room Numbers                     0.0
 Age                              0.0
 Distance                         0.0
 Teachers                         0.0
 Proportion of Poor Population    0.0
 Airport                          0.0
 Number of Hospital Beds          2.0
 Number of Hotel Rooms            0.0
 Waterbody                        0.0
 Rainfall                         0.0
 Bus Terminal                     0.0
 Parks                            0.0
 Sold                             0.0
 dtype: float64,
 3)

In [13]:
# Find the Nature of the Column 
df_house_price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   Residential Area               506 non-null    float64
 2   Air Quality                    506 non-null    float64
 3   Room Numbers                   506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   Proportion of Poor Population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number of Hospital Beds        498 non-null    float64
 10  Number of Hotel Rooms          506 non-null    float64
 11  Waterbody                      506 non-null    object 
 12  Rainfall                       506 non-null    int

In [14]:
# This is a Categorical Column so we can replace the null with the mode
df_house_price.Waterbody.mode()


0    River
Name: Waterbody, dtype: object

In [15]:
# Replace the Null Values
df_house_price.Waterbody =  df_house_price.Waterbody.fillna('River')

In [16]:
round(df_house_price.isnull().sum() / df_house_price.shape[0] * 100),3

(Price                            0.0
 Residential Area                 0.0
 Air Quality                      0.0
 Room Numbers                     0.0
 Age                              0.0
 Distance                         0.0
 Teachers                         0.0
 Proportion of Poor Population    0.0
 Airport                          0.0
 Number of Hospital Beds          2.0
 Number of Hotel Rooms            0.0
 Waterbody                        0.0
 Rainfall                         0.0
 Bus Terminal                     0.0
 Parks                            0.0
 Sold                             0.0
 dtype: float64,
 3)

In [17]:
# In the Column 'Number of Hospital Beds' so we can drop the null values
df_house_price.dropna(inplace=True)

In [18]:
round(df_house_price.isnull().sum() / df_house_price.shape[0] * 100),3

(Price                            0.0
 Residential Area                 0.0
 Air Quality                      0.0
 Room Numbers                     0.0
 Age                              0.0
 Distance                         0.0
 Teachers                         0.0
 Proportion of Poor Population    0.0
 Airport                          0.0
 Number of Hospital Beds          0.0
 Number of Hotel Rooms            0.0
 Waterbody                        0.0
 Rainfall                         0.0
 Bus Terminal                     0.0
 Parks                            0.0
 Sold                             0.0
 dtype: float64,
 3)

# prepare data for machine learning

In [19]:
#separate the target variable and independent variable 
target = df_house_price['Sold']

In [27]:
#independent variable
df_Independent = df_house_price.drop('Sold',axis=1)

## seperate the categorical columns from independent variables

In [28]:
df_cat = df_Independent.select_dtypes(include = 'object')

In [29]:
df_cat.head()

Airport Waterbody Bus Terminal
0     YES     River          YES
1      NO      Lake          YES
2      NO      None          YES
3     YES      Lake          YES
4      NO      Lake          YES

In [40]:
# encode the categorical variables 
encoded = pd.get_dummies(df_cat)
encoded = encoded.reset_index()

In [41]:
encoded.head()

index  Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0      0           0            1               0                         0   
1      1           1            0               1                         0   
2      2           1            0               0                         0   
3      3           0            1               1                         0   
4      4           1            0               1                         0   

   Waterbody_None  Waterbody_River  Bus Terminal_YES  
0               0                1                 1  
1               0                0                 1  
2               1                0                 1  
3               0                0                 1  
4               0                0                 1

## scale the numerical data

In [42]:
df_num = df_Independent.select_dtypes(include = 'number')

In [43]:
from sklearn.preprocessing import StandardScaler

In [44]:
# Initiate the StandardScaler 
ss = StandardScaler()

In [45]:
# Transform the data
scale = ss.fit_transform(df_num)

In [46]:
print(scale)

[[ 0.15479189 -1.28244435 -0.13709359 ... -0.35211111 -1.30206786
  -0.47313205]
 [-0.10659165 -0.59005882 -0.73319492 ... -0.16601943  0.21519397
  -0.77522463]
 [ 1.32012686 -0.59005882 -0.73319492 ... 16.71034024 -0.10422957
  -0.81123666]
 ...
 [ 0.14390091  0.11687263  0.16527665 ... -0.16252832 -0.66322077
   0.30293265]
 [-0.06302773  0.11687263  0.16527665 ...  0.40379146  0.6144734
   0.59750894]
 [-0.38975716  0.11687263  0.16527665 ... -0.54943507  0.45476163
   0.56369225]]


In [47]:
df_scale = pd.DataFrame(scale,columns = df_num.columns)

In [48]:
df_scale = df_scale.reset_index()
df_scale

index     Price  Residential Area  Air Quality  Room Numbers       Age  \
0        0  0.154792         -1.282444    -0.137094      0.408975 -0.115644   
1        1 -0.106592         -0.590059    -0.733195      0.189974  0.369741   
2        2  1.320127         -0.590059    -0.733195      1.276446 -0.260905   
3        3  1.178544         -1.301354    -0.828226      1.010516 -0.802977   
4        4  1.483492         -1.301354    -0.828226      1.222407 -0.505369   
..     ...       ...               ...          ...           ...       ...   
493    493 -0.019464          0.116873     0.165277      0.434572  0.022531   
494    494 -0.215501          0.116873     0.165277     -0.238073  0.291796   
495    495  0.143901          0.116873     0.165277      0.979230  0.798438   
496    496 -0.063028          0.116873     0.165277      0.720411  0.738208   
497    497 -0.389757          0.116873     0.165277     -0.366061  0.437057   

     Distance  Teachers  Proportion of Poor Population  \
0    0.172845  1.462167                      -1.066368   
1    0.476100  0.304963                      -0.484136   
2    0.495053  0.304963                      -1.199330   
3    1.054179 -0.111631                      -1.351886   
4    1.030487 -0.111631                      -1.017382   
..        ...       ...                            ...   
493 -0.637413 -1.176258                      -0.409958   
494 -0.732180 -1.176258                      -0.492534   
495 -0.779564 -1.176258                      -0.973995   
496 -0.684797 -1.176258                      -0.856429   
497 -0.599506 -1.176258                      -0.660486   

     Number of Hospital Beds  Number of Hotel Rooms  Rainfall     Parks  
0                  -1.640299              -0.352111 -1.302068 -0.473132  
1                  -0.384875              -0.166019  0.215194 -0.775225  
2                  -0.342847              16.710340 -0.104230 -0.811237  
3                   0.927490              -0.337843  0.454762 -0.680402  
4                   0.626514              -0.333593  1.253320 -1.404723  
..                       ...                    ...       ...       ...  
493                 0.981720              -0.164805 -0.982644  0.155116  
494                -0.872945               0.022197 -1.541636  0.522886  
495                -1.641654              -0.162528 -0.663221  0.302933  
496                 0.027273               0.403791  0.614473  0.597509  
497                 1.613499              -0.549435  0.454762  0.563692  

[498 rows x 13 columns]

In [53]:
X = pd.concat([df_scale,encoded], axis = 1)
X.head()

index     Price  Residential Area  Air Quality  Room Numbers       Age  \
0      0  0.154792         -1.282444    -0.137094      0.408975 -0.115644   
1      1 -0.106592         -0.590059    -0.733195      0.189974  0.369741   
2      2  1.320127         -0.590059    -0.733195      1.276446 -0.260905   
3      3  1.178544         -1.301354    -0.828226      1.010516 -0.802977   
4      4  1.483492         -1.301354    -0.828226      1.222407 -0.505369   

   Distance  Teachers  Proportion of Poor Population  Number of Hospital Beds  \
0  0.172845  1.462167                      -1.066368                -1.640299   
1  0.476100  0.304963                      -0.484136                -0.384875   
2  0.495053  0.304963                      -1.199330                -0.342847   
3  1.054179 -0.111631                      -1.351886                 0.927490   
4  1.030487 -0.111631                      -1.017382                 0.626514   

   ...  Rainfall     Parks  index  Airport_NO  Airport_YES  Waterbody_Lake  \
0  ... -1.302068 -0.473132      0           0            1               0   
1  ...  0.215194 -0.775225      1           1            0               1   
2  ... -0.104230 -0.811237      2           1            0               0   
3  ...  0.454762 -0.680402      3           0            1               1   
4  ...  1.253320 -1.404723      4           1            0               1   

   Waterbody_Lake and River  Waterbody_None  Waterbody_River  Bus Terminal_YES  
0                         0               0                1                 1  
1                         0               0                0                 1  
2                         0               1                0                 1  
3                         0               0                0                 1  
4                         0               0                0                 1  

[5 rows x 21 columns]

In [67]:
import statsmodels
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    target,
                                                    random_state=10,
                                                    test_size = 0.2)

In [68]:
X.head()

index     Price  Residential Area  Air Quality  Room Numbers       Age  \
0      0  0.154792         -1.282444    -0.137094      0.408975 -0.115644   
1      1 -0.106592         -0.590059    -0.733195      0.189974  0.369741   
2      2  1.320127         -0.590059    -0.733195      1.276446 -0.260905   
3      3  1.178544         -1.301354    -0.828226      1.010516 -0.802977   
4      4  1.483492         -1.301354    -0.828226      1.222407 -0.505369   

   Distance  Teachers  Proportion of Poor Population  Number of Hospital Beds  \
0  0.172845  1.462167                      -1.066368                -1.640299   
1  0.476100  0.304963                      -0.484136                -0.384875   
2  0.495053  0.304963                      -1.199330                -0.342847   
3  1.054179 -0.111631                      -1.351886                 0.927490   
4  1.030487 -0.111631                      -1.017382                 0.626514   

   ...  Rainfall     Parks  index  Airport_NO  Airport_YES  Waterbody_Lake  \
0  ... -1.302068 -0.473132      0           0            1               0   
1  ...  0.215194 -0.775225      1           1            0               1   
2  ... -0.104230 -0.811237      2           1            0               0   
3  ...  0.454762 -0.680402      3           0            1               1   
4  ...  1.253320 -1.404723      4           1            0               1   

   Waterbody_Lake and River  Waterbody_None  Waterbody_River  Bus Terminal_YES  
0                         0               0                1                 1  
1                         0               0                0                 1  
2                         0               1                0                 1  
3                         0               0                0                 1  
4                         0               0                0                 1  

[5 rows x 21 columns]

# decision tree classifier

In [76]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
tree_model = DecisionTreeClassifier()
tree_model.fit(X_train,y_train)
predictions = tree_model.predict(X_test)

#calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print("accuracy:",accuracy)

accuracy: 0.6
